modules/data-wrangling/index.qmd

```diff
@@ -165,11 +168,301 @@ For the expression field, there are many functions and operators that are useful
+
+```{r}
+#| eval: false
+penguins |>
+    mutate(sex = ifelse(sex == "male", "M", "F"))
+```
+
+```{r}
+#| echo: false
+penguins |>
+    mutate(sex = ifelse(sex == "male", "M", "F")) |>
+    select(species, sex) |>
+    head()
+```
+
+::: {.callout-caution}
+## Your turn!
+Let said that you think that the body mass of the penguins should be expressed as kilogram instead of gram. How could you create a **new** variable? Make sure you name it as `body_mass_kg`.
+
+Hint: The conversion from gram to kilogram can be expressed in a forms of $kg = g/1000$
+:::
+
+```{webr-teachr}
+# library the packages needed
+library(palmerpenguins)
+library(dplyr)
+
+# Select the column of interests
+penguins |>
+    <<mutate(body_mass_kg = body_mass_g/1000)>>
+
+???
+
+c(
+    "Did you use the right function?" != search_ast(.code, .fn = mutate),
+    "Did you provide the right expression?" != search_ast(.code, .expr = body_mass_g/1000)
+)
+```
+
+
+## group_by()
+
+It allows you to take the existing data and group the specific variables together for future operations. For example, with penguins data, if you want to rank penguins by weights within each species, instead of splitting the dataset by species, rank them individually and then rejoin the data. You can use the `group_by` function together with `mutate` to achieve the same operation.
+
+```{r}
+#| eval: false
+penguins |>
+    group_by(species) |>
+    mutate(ranking = rank(body_mass_g, ties.method = "min")) |>
+    ungroup()
+```
+
+```{r}
+penguins |>
+    group_by(species) |>
+    mutate(ranking = rank(body_mass_g, ties.method = "min")) |>
+    ungroup() |>
+    select(species, body_mass_g, ranking) |>
+    head()
+```
+
+:::{.callout-note}
+It is a good practice to use the `ungroup` function after you have done the operations. If you have not **ungrouped** the data, whatever operations you do after will take into account the grouping whether you want it or not.
+:::
+
+
+## summarize()
+As the name mentioned, it summarizes data into a new data frame based on the groupings and different arguments you included. The summary statistics are something you have to create very often. Here is an example of the summary statistics based on the penguin species.
+
+```{r}
+penguins |>
+    group_by(species) |>
+    summarize(
+        Average_body_mass = mean(body_mass_g, na.rm = TRUE),
+        Median_body_mass = median(body_mass_g, na.rm = TRUE),
+        SD_body_mass = sd(body_mass_g, na.rm = TRUE))
+```
+
+:::{.callout-note}
+The `na.rm` argument in the function is for ignoring the **NAs** values when calculating each statistic.
+:::
+
+:::{.callout-tip}
+## Did you know
+
+You may have noticed that we use the letter **z** in words like **summarize**. In British English, you might be more familiar with spelling it as **summarise**. In tidyverse, you can use either spelling when calling a function. So, whether you're used to writing summarize or summarising, the functionality is still the same either way.
```

Comment:
A bit wordy and possibly unimportant. Users will accidentally type these things the way they're familiar with and it won't matter.